### ISSUE TITLE
Support Host Network (--network host) mode on Docker Desktop for Mac and Windows

### ISSUE THREAD

**Issue by hexcowboy** [docker/roadmap#238]
## **Tell us about your request**
Right now, `docker run --network host <image>` is only available to Linux docker hosts. From the [official documentation](https://docs.docker.com/network/host/):
> The host networking driver only works on Linux hosts, and is not supported on Docker Desktop for Mac, Docker Desktop for Windows, or Docker EE for Windows Server.

## **Which service(s) is this request for?**
Docker Desktop for Mac, Docker Desktop for Windows

## **Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Making a docker container with dynamic port mappings is not possible right now.

For example, a container that allows a user to accept connections on arbitrary ports through a netcat listener. During a network penetration test, the user may be testing outbound firewall rules and needs to dynamically accept connections on different ports in the container, eg `53`, `80`, `443`, `1337`, `9001`, or any other *arbitary* port. The user will not know these ports before starting the container, so it's not possible for the user to run the container with `docker run -p <ports> <image>`.

With `docker run --network=host`, this could be accomplished by just using the same network schema as `localhost`.

## **Are you currently working around the issue?**
The "best" workaround I know of would to add an `EXPOSE 1-1000` in the Dockerfile and run the container with `docker run -P <image>`, which is very slow and does still limits the range of open ports. Using the full port range of `EXPOSE 1-65535` hangs my terminal, so I don't know how effective it is.

The `-P` flag is documented here: https://docs.docker.com/network/links/#connect-using-network-port-mapping

This is not a sustainable workaround.

## **Additional context**

- https://stackoverflow.com/questions/49323225/expose-all-ports-for-a-docker-image/49323975
- https://stackoverflow.com/questions/54165483/docker-alternative-to-network-host-on-macos-and-windows

**Comment by Amondale** [comment]
Although documentation stats not supported on any O/S but Linux, it also doesn't seem to work correctly (although it doesn't throw error) on WSL2. The swarm configurations likewise don't function as they do in "bare metal" Linux, which can make for a *ton* of confusion in the docs.

There is a large and growing population of Mac/Windows users that are not fully supported by the extant documentation, and issues like this (many other networking issues exist on those platforms too).

I plan to follow this thread closely, and would welcome the chance to help refine docs so they apply across the board, or exceptions are noted where they exist.

**Comment by gramss** [comment]
What is basically needed to make this work? Where would one need to start looking to make this work on MacOS?
Any basic information on how to start that would be much appreciated.
I think technically this should be doable, also from the community. It's currently just not on the roadmap..?

**Comment by zfil** [comment]
Having this supported could maybe help with the horrible network performance I see on macos docker ...

This makes for instance using postgres  in a container very painful.
Running our test suite using a postgres in a container and with docker 4.6:
    virtualization.framework: 16min
    hypervisor.framework: 10min
    locally installed postgres: 3min

**Comment by Amondale** [comment]
I'd like to see the memory footprint reduced in the Windows version, if at all possible. If it was possible to run without Docker Desktop as it is in Linux, I bet we'd see some reduction in the 2+ GB that Desktop (vmem process) consumes with no running containers.

**Comment by christophermclellan** [comment]
Hi - thanks for this issue and sorry for the delayed reply. We'll do some scoping on this and circle back on this thread shortly.

**Comment by wiredwiz** [comment]
This is certainly a feature I would love to see support for.  lacking host network support makes it impossible for me to run services that need to be network aware on anything other than linux, since otherwise they only see the network bridge.

**Comment by hexcowboy** [comment]
Another limitation I've found - not being able to run a local Redis cluster https://redis.io/docs/manual/scaling/#redis-cluster-and-docker

**Comment by ReubenFrankel** [comment]
Related to https://github.com/docker/desktop-linux/issues/87

**Comment by gramss** [comment]
Now with the [new release 4.14.0](https://docs.docker.com/desktop/release-notes/#4140) of Docker Desktop on Mac utilizing the Apple Virtualization Framework, it should be possible to add another network adapter featuring the bridged network aka. network_mode = host ?

https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656724-networkdevices

Is this a big challenge? Can somebody point out the hypervisor configuration for the Apple Virtualization Framework here? :)

Also thanks for putting it in the "Considering" space of the docker-roadmap ! 👍

**Comment by logical-and** [comment]
+1 - would be really nice to get this implemented as it works on Linux!

**Comment by jwoodrow** [comment]
> Now with the [new release 4.14.0](https://docs.docker.com/desktop/release-notes/#4140) of Docker Desktop on Mac utilizing the Apple Virtualization Framework, it should be possible to add another network adapter featuring the bridged network aka. network_mode = host ?
> 
> [developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656724-networkdevices](https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/3656724-networkdevices)
> 
> Is this a big challenge? Can somebody point out the hypervisor configuration for the Apple Virtualization Framework here? :)
> 
> Also thanks for putting it in the "Considering" space of the docker-roadmap ! 👍

Is there any way to get more updates on this ? I think this would be an extremely appreciated feature for mac users

**Comment by davidthornton** [comment]
Here's the issue link to follow it through the roadmap!

https://github.com/docker/roadmap/issues/238

**Comment by davidthornton** [comment]
Here's the link to the filter to track it's progress!

https://github.com/docker/roadmap/projects/1?card_filter_query=network+host

**Comment by dewdrinker19** [comment]
Yes please.  Very much please.

**Comment by JCHH2** [comment]
> Here's the link to the filter to track its progress!
> 
> https://github.com/docker/roadmap/projects/1?card_filter_query=network+host

It doesn't seem to be getting traction or is it too early to tell?

**Comment by djirik** [comment]
I've been waiting for 5 years for that. Can you please do it? If the security is main concern please disable this feature by default.

**Comment by podjackel** [comment]
bump. I have several tools that I would love to be able to containerize that need layer2 access.

**Comment by ShutdownRepo** [comment]
For everyone's bummed about this lack, the solution may very well be to switch to a more active project 🤷 It's been five years. For instance, https://orbstack.dev/, while not open-source, seems very promising and already supporting the `--network=host` mode (among other things). I'm amazed to see independent projects like this one surpass Docker Desktop (for mac in this case) performance and feature-wise.

**Comment by JCHH2** [comment]
> https://orbstack.dev/

Website has been registered for only 1 week.  How well do you know it?

**Comment by ShutdownRepo** [comment]
> > https://orbstack.dev/
> 
> Website has been registered for only 1 week. How well do you know it?

Not well myself, but I work with 2 people that tested it and that raised excellent feedback.

**Comment by reececomo** [comment]
Any update on this? Host networking is a huge enabler for local development of WebRTC apps & games.

**Comment by jupadev** [comment]
it would be great have this flag for mac and windows in order to run some container against host machine for early testing

**Comment by dontcrash** [comment]
+1 to this

**Comment by jwoodrow** [comment]
> > > https://orbstack.dev/
> 
> > 
> 
> > Website has been registered for only 1 week. How well do you know it?
> 
> 
> 
> Not well myself, but I work with 2 people that tested it and that raised excellent feedback.

I've been using orbstack for the past 2 months and have had 0 complaints apart from details like the fact I can't access logs from the gui or images. It's faster, lighter and has more important features than docker desktop ironically.

Docker has been sitting on this issue for months among many other issues and I just feel like they don't care at this point...

**Comment by dontcrash** [comment]
OrbStack has issues as well, I tried to run a Traefik container and bind port 53, no matter what I did, it would not bind. Other ports were fine.

**Comment by kdrag0n** [comment]
> OrbStack has issues as well, I tried to run a Traefik container and bind port 53, no matter what I did, it would not bind. Other ports were fine.

@dontcrash This is probably because you already have a service bound to port 53 on macOS. To identify the process:

```bash
lsof -Pwni | grep :53
```

Let me know if you find anything!

**Comment by dontcrash** [comment]
> > OrbStack has issues as well, I tried to run a Traefik container and bind port 53, no matter what I did, it would not bind. Other ports were fine.
> 
> @dontcrash This is probably because you already have a service bound to port 53 on macOS. To identify the process:
> 
> ```shell
> lsof -Pwni | grep :53
> ```
> 
> Let me know if you find anything!

Binds to OrbStack yes, but it does not work. I have Traefik hosting a UDP tunnel to a DNS server, the request hits OrbStack and nothing further, does not go to the container, another issue I am seeing is that even with other services the source IP is 127.0.0.1, this is not how host networking should behave.
This problem would be solved if the Docker team would push this from considering after several years and just do it.

**Comment by chrisdoherty4** [comment]
Lets try and keep the conversation focused on the issue (as opposed to OrbStack troubleshooting).

**Comment by kcognata** [comment]
Would love to see this feature. It's been a linux exclusive for far too long.

**Comment by RedNo7** [comment]
+1 for this.

Orbstack does not have any timing for this either; feedback was there is not enough support to prioritise it.
https://github.com/orbstack/orbstack/issues/183

**Comment by leosamuele221** [comment]
+1
Cant see client IP on traefik

**Comment by nrjohnstone** [comment]
Would this also restore the old functionality we used to have with the NAT router way back in 19.x that allowed us to route traffic from our windows host to static IPs on containers? That was excellent functionality that allows you to mimic a deployed environment plus have routable host names in your localhosts file. Having to port map everything and access it via "127.0.0.1" is a very poor substitute, especially if you have quite  a number of dependencies you need running in containers locally.

**Comment by thaJeztah** [comment]
In host-mode networking, the container does not have a networking namespace, and as such doesn't have an IP-address.

**Comment by sourabhjaiswal** [comment]
+1. Would love to see this feature as in some cases we need to run a linux container on windows system and want the IP of host to be used inside container.

**Comment by varzantr** [comment]
+1 facing the same issue. Found [this workaround](https://medium.com/@lailadahi/getting-around-dockers-host-network-limitation-on-mac-9e4e6bfee44b) on Medium but for me it doesn't work.

**Comment by enzofrnt** [comment]
I'd like to see the inclusion of this feature, and I also have other recommendations to mention.

Is it possible to add the ability to bind the interface from LinuxKit to MacOS concurrently ? This would facilitate the utilization of IPVLAN and MACVLAN with an operational parent interface, or any other mechanism that would allow us to use these types of networks and expose IPs on the local network.

There's a workaround that worked previously on x86, but not on Apple Silicon Mac :
[docker-tuntap-osx](https://github.com/AlmirKadric-Published/docker-tuntap-osx)

There is a big lack of networking capabilities on the Mac Version of Docker.

**Comment by danielfinke** [comment]
@enzofrnt I am using [docker-mac-net-connect](https://github.com/chipmk/docker-mac-net-connect) as a workaround on my M1 Mac. I have a container that listens on a wide port range so I need to route traffic to it directly.

**Comment by enzofrnt** [comment]
Hi, @danielfinke thanks !
I will try it because I find nothing else, but that will be way better to has it implemented directly in docker…

**Comment by enzofrnt** [comment]
@danielfinke i'm that will make me able to do make MacVlan or ipvlan directly to my local network.
As my computer is 192.168.1.45 i want container in this range and available for other devices.. Any help there ?

**Comment by danielfinke** [comment]
@enzofrnt It only gets you as far as `docker-tuntap-osx` in that you can route traffic from the host directly to containers. I think you would need to add a proxy on top if you wanted to use the host's IP, but I am not sure.

**Comment by enzofrnt** [comment]
@danielfinke i'm currently working on a Mac with M1 chip and they use linux kit instead of HyperKit.. 
So the tuntap thing didn't work for me.
Maybe with some modifications that will work but i'm not able do this kind of modifications.
So, i'm juste in front of a wall.

**Comment by enzofrnt** [comment]
@danielfinke there is what I wonder : 
![image](https://github.com/docker/roadmap/assets/63660254/66e27976-fe55-4a39-a44c-1d5df288bb2e)

**Comment by nikjuice** [comment]
+1

**Comment by miamilabs** [comment]
I was trying to run home assistant in host mode which did not work for me. For my understanding the mdns traffic was  not forwarded correctly. 
From the reply which i got its because of the host mode is missing on mac os.

**Comment by quiquelhappy** [comment]
wen windows

**Comment by outsideryn** [comment]
+1

**Comment by vadimen** [comment]
any ideas how to do this [[Bind dedicated ip to docker container in Windows Server 2022](https://devops.stackexchange.com/questions/19117/bind-dedicated-ip-to-docker-container-in-windows-server-2022)](https://devops.stackexchange.com/questions/19117/bind-dedicated-ip-to-docker-container-in-windows-server-2022)

**Comment by 4-FLOSS-Free-Libre-Open-Source-Software** [comment]
Version Docker Desktop 4.29.0
Settings ⇒ Features in development ⇒ Enable host networking
Host networking allows containers that are started with --net=host to use localhost to connect to TCP and UDP services on the host. It will automatically allow software on the host to use localhost to connect to TCP and UDP services in the container. Sign in required.

![Enable host networking](https://github.com/docker/roadmap/assets/46166740/7ed61aec-690e-47e0-8ebb-9539c1f72104)

**Comment by vadimen** [comment]
> Version Docker Desktop 4.29.0 Settings ⇒ Features in development ⇒ Enable host networking Host networking allows containers that are started with --net=host to use localhost to connect to TCP and UDP services on the host. It will automatically allow software on the host to use localhost to connect to TCP and UDP services in the container. Sign in required.
> 
> ![Enable host networking](https://private-user-images.githubusercontent.com/46166740/320818234-7ed61aec-690e-47e0-8ebb-9539c1f72104.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NjA2MzcsIm5iZiI6MTcxMjY2MDMzNywicGF0aCI6Ii80NjE2Njc0MC8zMjA4MTgyMzQtN2VkNjFhZWMtNjkwZS00N2UwLThlYmItOTUzOWMxZjcyMTA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDEwNTg1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNGE1NWYxZDY2OGZhZDI0ZGFhMDNlYzY1MzkwOTBkZjZiNWY2MjUzMTZkZWNkMzJmZjFjY2Q1MzQ2ZDE2YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eo7RHM08mBGSyu43pIlJxI86nRpuw4HcJnrWJo6MHRU)

that is very cool, can we activate it using powershell on windows server core ?

**Comment by 4-FLOSS-Free-Libre-Open-Source-Software** [comment]
C:\Users\$env:USERNAME\AppData\Roaming\Docker\settings.json

`"hostNetworkingEnabled": true,`

> > Version Docker Desktop 4.29.0 Settings ⇒ Features in development ⇒ Enable host networking Host networking allows containers that are started with --net=host to use localhost to connect to TCP and UDP services on the host. It will automatically allow software on the host to use localhost to connect to TCP and UDP services in the container. Sign in required.
> > ![Enable host networking](https://private-user-images.githubusercontent.com/46166740/320818234-7ed61aec-690e-47e0-8ebb-9539c1f72104.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NjA2MzcsIm5iZiI6MTcxMjY2MDMzNywicGF0aCI6Ii80NjE2Njc0MC8zMjA4MTgyMzQtN2VkNjFhZWMtNjkwZS00N2UwLThlYmItOTUzOWMxZjcyMTA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDEwNTg1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNGE1NWYxZDY2OGZhZDI0ZGFhMDNlYzY1MzkwOTBkZjZiNWY2MjUzMTZkZWNkMzJmZjFjY2Q1MzQ2ZDE2YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eo7RHM08mBGSyu43pIlJxI86nRpuw4HcJnrWJo6MHRU)
> 
> that is very cool, can we activate it using powershell on windows server core ?

Try my script:

```powershell
# Read the JSON file
$path = 'C:\Users\{0}\AppData\Roaming\Docker\settings.json' -f $env:USERNAME
$json = Get-Content -Raw -Path $path | ConvertFrom-Json

# Check if the property exists
if ($json.PSObject.Properties.Name -notcontains 'hostNetworkingEnabled') {
    # Add the property if it does not exist
    $json | Add-Member -Type NoteProperty -Name hostNetworkingEnabled -Value $true
} else {
    # Change the value
    $json.hostNetworkingEnabled = $true
}

# Save the changes back to the file
$json | ConvertTo-Json -Depth 100 | Set-Content -Path $path

**Comment by enzofrnt** [comment]
this thing enable IPVLAN ?

**Comment by bryansh** [comment]
Good news!  This has been added as a beta feature in Docker Desktop 4.29.

**Comment by miamilabs** [comment]
> Good news! This has been added as a beta feature in Docker Desktop 4.29.

yea its amazing for some features. I managed to run home assistant now outside of VM. But faced now new issues with USB stuff which cant be mouted to docker since 6 years as well :)


Good Luck all. Hope people can resolve problems with this update

**Comment by ca0x** [comment]
Yeah, I also tried with scrypted and homebridge. It's not working as intended. Although it works with nginx. 

Here are the open ports in the VM and on the host with both nginx and homebridge when using ```network_mode: host```


### nginx

```
nginx (Docker VM)
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN  
tcp        0      0 :::80                   :::*                    LISTEN 

nginx (host)
com.docke  879            user  238u  IPv6 0x1234...      0t0    TCP *:80 (LISTEN)
```


### homebridge

```
homebridge (Docker VM)

tcp        0      0 :::8581                 :::*                    LISTEN
tcp        0      0 :::51826                :::*                    LISTEN
udp        0      0 0.0.0.0:62891           0.0.0.0:*

homebridge (host)

com.docke  879            user  289u  IPv6 0x1234...     0t0    UDP *:62891
```

It seems that only the ports listening on IPv4 addresses are exposed.

**Comment by vadimen** [comment]
> C:\Users$env:USERNAME\AppData\Roaming\Docker\settings.json
> 
> `"hostNetworkingEnabled": true,`
> 
> > > Version Docker Desktop 4.29.0 Settings ⇒ Features in development ⇒ Enable host networking Host networking allows containers that are started with --net=host to use localhost to connect to TCP and UDP services on the host. It will automatically allow software on the host to use localhost to connect to TCP and UDP services in the container. Sign in required.
> > > ![Enable host networking](https://private-user-images.githubusercontent.com/46166740/320818234-7ed61aec-690e-47e0-8ebb-9539c1f72104.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NjA2MzcsIm5iZiI6MTcxMjY2MDMzNywicGF0aCI6Ii80NjE2Njc0MC8zMjA4MTgyMzQtN2VkNjFhZWMtNjkwZS00N2UwLThlYmItOTUzOWMxZjcyMTA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDEwNTg1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNGE1NWYxZDY2OGZhZDI0ZGFhMDNlYzY1MzkwOTBkZjZiNWY2MjUzMTZkZWNkMzJmZjFjY2Q1MzQ2ZDE2YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eo7RHM08mBGSyu43pIlJxI86nRpuw4HcJnrWJo6MHRU)
> > 
> > 
> > that is very cool, can we activate it using powershell on windows server core ?
> 
> Try my script:
> 
> ```powershell
> # Read the JSON file
> $path = 'C:\Users\{0}\AppData\Roaming\Docker\settings.json' -f $env:USERNAME
> $json = Get-Content -Raw -Path $path | ConvertFrom-Json
> 
> # Check if the property exists
> if ($json.PSObject.Properties.Name -notcontains 'hostNetworkingEnabled') {
>     # Add the property if it does not exist
>     $json | Add-Member -Type NoteProperty -Name hostNetworkingEnabled -Value $true
> } else {
>     # Change the value
>     $json.hostNetworkingEnabled = $true
> }
> 
> # Save the changes back to the file
> $json | ConvertTo-Json -Depth 100 | Set-Content -Path $path
> ```

I have both "hostNetworkingEnabled": true and "allowExperimentalFeatures": true. Stopped and started docker service. Still I am getting "docker: Error response from daemon: network host not found." .

**Comment by ca0x** [comment]
> Yeah, I also tried with scrypted and homebridge. It's not working as intended. Although it works with nginx.
> 
> Here are the open ports in the VM and on the host with both nginx and homebridge when using `network_mode: host`
> 
> ### nginx
> ```
> nginx (Docker VM)
> tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN  
> tcp        0      0 :::80                   :::*                    LISTEN 
> 
> nginx (host)
> com.docke  879            user  238u  IPv6 0x1234...      0t0    TCP *:80 (LISTEN)
> ```
> 
> ### homebridge
> ```
> homebridge (Docker VM)
> 
> tcp        0      0 :::8581                 :::*                    LISTEN
> tcp        0      0 :::51826                :::*                    LISTEN
> udp        0      0 0.0.0.0:62891           0.0.0.0:*
> 
> homebridge (host)
> 
> com.docke  879            user  289u  IPv6 0x1234...     0t0    UDP *:62891
> ```
> 
> It seems that only the ports listening on IPv4 addresses are exposed.

Disabling IPv6 in the VM solved the ports issue, exposing all ports and allowing network access to the Homebridge UI. However, it appears that mDNS is not functioning, and the service remains undiscoverable by HomeKit.


### Access VM
```
docker run -it --rm --privileged --pid=host justincormack/nsenter1
```

### Disable IPv6
```
sysctl -w net.ipv6.conf.all.disable_ipv6=1
```

**Comment by akerouanton** [comment]
@ca0x Thanks for reporting. It seems you're hitting two limitations:

Software working in dual-stack mode, or IPv6-only can't benefit from this improvement. As we're working toward adding better support for IPv6, we're going to fix this mishap soon -- most probably in v4.31.

mDNS is currently not supported. We need to investigate exactly why.

@vadimen Are you trying to use this feature with Windows Containers? This improvement works only if there's a Linux VM running.

@Nemesis7 What ports Scrypted is listening to? Are some of them bound by DD on the host? It's probably worth filling a new issue in either https://github.com/docker/for-win or https://github.com/docker/for-mac.

**Comment by vadimen** [comment]
> @ca0x Thanks for reporting. It seems you're hitting two limitations:
> 
> Software working in dual-stack mode, or IPv6-only can't benefit from this improvement. As we're working toward adding better support for IPv6, we're going to fix this mishap soon -- most probably in v4.31.
> 
> mDNS is currently not supported. We need to investigate exactly why.
> 
> @vadimen Are you trying to use this feature with Windows Containers? This improvement works only if there's a Linux VM running.
> 
> @Nemesis7 What ports Scrypted is listening to? Are some of them bound by DD on the host? It's probably worth filling a new issue in either https://github.com/docker/for-win or https://github.com/docker/for-mac.

Yes, I was trying to use it with windows containers. Could you please suggest me any solution for this use case https://devops.stackexchange.com/questions/19117/bind-dedicated-ip-to-docker-container-in-windows-server-2022  ?

**Comment by akerouanton** [comment]
@vadimen I'm not super familiar with Windows Containers networking. I think it's worth looking for a similar issue or asking on https://github.com/microsoft/Windows-Containers.

**Comment by Nemesis7** [comment]
> @Nemesis7 What ports Scrypted is listening to? Are some of them bound by DD on the host? It's probably worth filling a new issue in either https://github.com/docker/for-win or https://github.com/docker/for-mac.

By default it's port 10443, plus there one port for HomeKit which is variable and there is the mDNS port 5353, but shouldn't this change allow any container with `network_mode: host` to be reachable via `localhost`? For _Scrypted_ I can create a new issue, but I think reachability via `localhost` is related to this thread, no ?

**Comment by akerouanton** [comment]
@Nemesis7 It's definitely related but it'd be better to have a proper bug report.

**Comment by AndreKR** [comment]
Hm, am I missing something?

I'm running 4.30.0:

![image](https://github.com/docker/roadmap/assets/1188538/f98ff793-6818-416d-a6d7-a3e3944f5fe1)

Under "Features in development" I enabled "Enable host networking":

![image](https://github.com/docker/roadmap/assets/1188538/076e9449-a6a5-4adc-b750-5024164b8770)

(For unknown reasons that checkbox only became enabled when I logged into an account, it's unclear to me how networking and login are related.)

If I now run netcat in listen mode in a container:

![image](https://github.com/docker/roadmap/assets/1188538/c2885f6e-9be2-473d-8d41-19855975d087)

I am unable to connect to that from outside the container:

![image](https://github.com/docker/roadmap/assets/1188538/a85b8afe-45aa-4b15-bc89-21c317b7be92)

But if I exec into the container and try the connection from there, it works:

![image](https://github.com/docker/roadmap/assets/1188538/385b7efc-48e4-4daf-a995-d7e98d609e31)

That should work, right?

**Comment by tangledhelix** [comment]
Using `-l` and `-p` together is an error. From `nc(1)`:

```
     -l      Used to specify that nc should listen for an incoming connection
             rather than initiate a connection to a remote host.  It is an
             error to use this option in conjunction with the -p, -s, or -z
             options.  Additionally, any timeouts specified with the -w option
             are ignored.

     -p source_port
             Specifies the source port nc should use, subject to privilege
             restrictions and availability.  It is an error to use this option
             in conjunction with the -l option.
```

Also, when using listener mode the port is random:

```
/ # nc -lv
listening on [::]:56323 ...
```

If you run a listener like this, and then try to connect to it using localhost on the outer system, it still won't work, because you can see it is bound to `[::]` which is the IPv6 stack. The `--net=host` parameter will only bind to IPv4, so the last piece of the puzzle is to specify an IPv4 address to bind to...

```
/ # nc -lv 0.0.0.0
listening on 0.0.0.0:59271 ...
```

And after this, connecting to localhost on the outer host, using the port in the message (59271 in this example) should work. I tested this on Mac, although I see you're using Windows. I would hope it works the same, but YMMV.

**Comment by AndreKR** [comment]
What image are you using for your container? The netcat syntax you're using is for BSD netcat, for GNU and Busybox netcat my syntax should be ok.

I tried installing BSD netcat in my Alpine container but I couldn't get it to work with your commands:

![image](https://github.com/docker/roadmap/assets/1188538/a66465eb-fa22-4bdf-b97c-4651f8ec8b97)

I tried specifying a local IPv4 address to rule out the IPv6 issue you mention with `nc -l -p 1234 -s 0.0.0.0` and `nc -l -p 1234 -s 127.0.0.1` but the netcat is still not reachable from the host.

**Comment by tangledhelix** [comment]
Based on your screenshot, to replicate your environment, I used alpine:latest.

```
❯ docker image ls alpine:latest
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
alpine       latest    8e1d7573f448   2 years ago   5.33MB
```

I will admit I probably used the manpage from macOS (which would have the BSD version), not realizing the disparity between it and GNU/Linux. But the behavior I tested and pasted output from came from `alpine:latest`:

```
❯ docker run -it --rm --net=host alpine:latest
/ # nc -lv 0.0.0.0
listening on 0.0.0.0:55201 ...
```

I see now that your syntax works and has the right port (which it clarifies with `-v`). But, it binds to IPv6 only, which leads back to my observation about needing to use IPv4:

```
/ # nc -l -p 1234 -v
listening on [::]:1234 ...
```

If you bind to the IPv4 stack it then allows connection from the outside host:

```
/ # nc -lk -p 1234 -v -s 0.0.0.0 -e date
listening on 0.0.0.0:1234 ...
connect to 192.168.65.3:1234 from 192.168.65.1:38201 (192.168.65.1:38201)
```

```
❯ nc localhost 1234
Sat May 18 22:32:06 UTC 2024
```

**Comment by AndreKR** [comment]
Using your command line doesn't work for me:

![image](https://github.com/docker/roadmap/assets/1188538/041af100-1c7f-4a07-98f8-220e67b97965)

**Comment by tangledhelix** [comment]
Ah, sorry to see that. Unfortunately I have no explanation for it. I would guess it is a difference in the Mac vs. Windows environment (either the network stack, or the way Docker Desktop works with the host OS, or both). I don't have any Windows machines around here to test it.

**Comment by s13n** [comment]
It doesn't work with me. I'm using v4.33.0 on Windows 11, and have enabled the feature after signing in. Alas, `docker info` doesn't show the host network mode as available, and starting a container with `--net=host` throws an error.

`  Network: ics internal l2bridge l2tunnel nat null overlay private transparent`

Am I doing something wrong? Do I have to create the host network somehow?

**Comment by akerouanton** [comment]
@s13n This feature isn't compatible with Windows containers. It's meant to be used with Linux containers, on both macOS and Windows.

**Comment by s13n** [comment]
@akerouanton Thanks! That wasn't clear to me, and the explanatory text near the tick box in the settings dialog doesn't mention it. Maybe it would be an idea to disable the choice when Docker Desktop is in Windows container mode.

**Comment by akerouanton** [comment]
@s13n Thanks for reporting! This should be fixed in the next release (4.34).
